feat(one-use-login-link): Changes login links to single use

sbinlondon opened the pull request:
Addresses opencollective/opencollective#2192

![single-use-login](https://user-images.githubusercontent.com/37520401/66394435-08664b80-e9cd-11e9-94ee-5a8d18fd4a88.gif)

This creates single-use login links by adding a field to the first JWT token that is processed when a user signs in.

Hopefully I've updated all the relevant database and GraphQL bits, but let me know if I've missed anything.

Feedback and questions welcome 👍

sbinlondon commented:
Can't seem to request a review (since I don't have write access) so tagging @znarf 😄

znarf commented on migrations/20191007173134-add-lastlogin-column-to-users.js:
```diff
@@ -0,0 +1,13 @@
+  down: function(queryInterface) {
+    return queryInterface.removeColumn('Users', 'lastLogin');
+  },
+};
```
Have you considered naming the column `lastLoginAt`?

That would match the convention we have for Date columns.

sbinlondon replied:
That makes more sense. I'll be sure to update that.

znarf commented on .forestadmin-schema.json:
```diff
@@ -5287,6 +5287,21 @@
+          "reference": null,
+          "inverseOf": null,
+          "validations": []
+        },
```
I'm curious how you got that modified? Was it done automatically?

This file is a bit new for us, and we're still trying to figure the best practices when modifying it.

sbinlondon replied:
I did it myself. I was trying to figure out how to make sure my new `lastLogin` column was included everywhere it ought to be and thought that this needed to be modified as well.

znarf replied:
No Forest API keys were required for that?

sbinlondon replied:
No - i just edited the file. It didn't fuss or anything 🤔

znarf replied:
Ok, I get it!

znarf replied:
I introduced yesterday the first version of a tool to update that file. https://github.com/opencollective/opencollective-api/pull/2694 It's not documented yet and this is why I was interested.

znarf commented on server/middleware/security/authentication.js:
```diff
@@ -84,9 +84,24 @@ export const checkJwtExpiry = (req, res, next) => {
 export const _authenticateUserByJwt = (req, res, next) => {
   if (!req.jwtPayload) return next();
   const userid = Number(req.jwtPayload.sub);
+  const lastLogin = req.jwtPayload.last;
```
You set `lastLogin` here but are still using `req.jwtPayload.last` later. Maybe you should try to always use `lastLogin` once it's set?

sbinlondon replied:
Agreed! I thought for some reason that perhaps we wouldn't want to be obvious about the naming of the last login timestamp in a JWT, since they can be decoded, but that's probably just me overthinking 😅

znarf commented on server/middleware/security/authentication.js:
```diff
@@ -84,9 +84,24 @@ export const checkJwtExpiry = (req, res, next) => {
+          const now = new Date();
+          user.update({ lastLogin: now });
+        }
+      }
```
Some code comments would help understand this section. I also think it there is some potential to make it easier to read.

znarf replied:
For example, one explanation before the `if` and `else` conditions could be interesting.

znarf replied:
It's now already super clear but I'm wondering if this could be simplified.

I see that in both cases we do the following:

```
const now = new Date();
user.update({ lastLoginAt: now });
```

Maybe we could have that only once by moving it out of the conditions. I can imagine this should then allow some further simplification.

sbinlondon replied:
I see! Really good point. I'll update that now 😊

sbinlondon replied:
Updated the code. Let me know when you want to chat through it or if you think it needs some more changes!

sbinlondon commented on migrations/05-user-model.js:
```diff
@@ -60,6 +60,9 @@ module.exports = {
       seenAt: DataTypes.DATE,
+      lastLogin: {
+        type: DataTypes.DATE,
+      },
```
Ah, good point - sorry about that! I'll be sure to remove it.

sbinlondon commented:
I've updated the PR based on your comments @znarf. I hope the comments in the code makes it better to read; I've also renamed some of the variables to hopefully make things more clear.

znarf commented:
Looks good for me, asking a second review from @Betree

Betree commented on server/middleware/security/authentication.js:
```diff
@@ -84,9 +84,33 @@ export const checkJwtExpiry = (req, res, next) => {
+          const lastLoginInDb = user.lastLoginAt;
+          if (lastLoginInJwt.getTime() !== lastLoginInDb.getTime()) {
+            throw errors.Unauthorized('Invalid login link - multiple uses');
+          }
```
Can't `user.lastLoginAt` be `null`? (it would crash in the condition below)

sbinlondon replied:
You're right, when the db initialises, then `lastLoginAt` will be null. But it shouldn't crash in the `if` statement because of the code that decides whether you go into the `if` statement or not (line 104).

If it's the user's first login, i.e. they just signed up, then line 104 takes them around the `if` statement and straight into the bit where we overwrite the `lastLoginAt` in the db with the first login time. (This is because both of the conditions are falsy.)

So in all other cases, `user.lastloginAt` should actually be a timestamp, meaning we can: 1) prevent re-use of the first login token if `req.jwtPayload.lastLoginAt !== 'firstLogin'` evaluates to true; and 2) takes us into the `if` statement if it isn't the user's first login, allowing us to check that the `lastLoginAt` in the JWT matches that in the db.

Betree replied:
Alright, thanks for the details
